src/OtcNexus.sol

```diff
@@ -0,0 +1,603 @@
+     * @notice take a fixed price RFS
+     * @param _id id of the RFS
+     * @param _paymentTokenAmount amount of tokens used to pay for the RFS
+     * @param index index of the token used to buy
+     * @return success
+     */
+    function takeFixedRfs(
+        uint256 _id,
+        uint256 _paymentTokenAmount,
+        uint256 index
+    ) external returns (bool success) {
+        return takeRfs(_id, _paymentTokenAmount, index, RfsType.FIXED);
+    }
+
+    /**
+     * @notice take a dynamic price RFS
+     * @param _id id of the RFS
+     * @param _paymentTokenAmount amount of tokens used to pay for the RFS
+     * @param index index of the token used to buy
+     * @return success
+     */
+    function takeDynamicRfs(
+        uint256 _id,
+        uint256 _paymentTokenAmount,
+        uint256 index
+    ) external returns (bool success) {
+        return takeRfs(_id, _paymentTokenAmount, index, RfsType.DYNAMIC);
+    }
+
+    /**
+     * @notice take a fixed price RFS
+     * @param _id id of the RFS
+     * @param _paymentTokenAmount amount of tokens used to pay for the RFS
+     * @param index index of the token used to buy
+     * @param expectedRfsType type of the Rfs: FIXED or DYNAMIC
+     * @return success
+     */
+    function takeRfs(
+        uint256 _id,
+        uint256 _paymentTokenAmount,
+        uint256 index,
+        RfsType expectedRfsType
+    ) internal returns (bool success) {
+        RFS memory rfs = getRfs(_id);
+
+        // sanity checks
+        if (rfs.typeRfs != expectedRfsType) revert OtcNexus__InvalidRfsType();
+        if (rfs.removed) revert OtcNexus__RfsRemoved();
+        if (_paymentTokenAmount == 0) revert OtcNexus__InvalidTokenAmount();
+        if (index >= rfs.tokensAccepted.length) revert OtcNexus__InvalidTokenIndex();
+        if (
+            IERC20(rfs.tokensAccepted[index]).allowance(msg.sender, address(this)) <
+            _paymentTokenAmount
+        ) revert OtcNexus__AllowanceToken1TooLow();
+
+        //0.25% fee on the payment token of the taker FLAT FEE
+        (uint256 feeAmount1, uint256 _paymentTokenAmountAfterTakerFee) = takeFee(
+            takerFee,
+            _paymentTokenAmount
+        );
+        // compute _amount0 (depends on rfs.typeRfs, so this is left to takeFixedRfs/takeDynamicRfs)
+        uint256 _amountBuying = computeAmount0(
+            _id,
+            _paymentTokenAmountAfterTakerFee,
+            index,
+            expectedRfsType
+        );
+        if (_amountBuying > rfs.amount0) revert OtcNexus__Amount0TooHigh();
+
+        //0.25% fee on the token0 of the maker FLAT FEE if DEPOSITED, otherwise 0.5% if APPROVED
+        (uint256 feeAmount2, uint256 _paymentTokenAmountAfterMakerFee) = takeFee(
+            rfs.interactionType == TokenInteractionType.TOKEN_DEPOSITED
+                ? makerFeeIfDeposited
+                : makerFeeIfNotDeposited,
+            _paymentTokenAmountAfterTakerFee
+        );
+        // transfer token1 to the maker
```

Comment:
1. sth fishy is going below - why are we splitting those transfer for two same transfers? It does not make sense, maybe the intention was different?


2. shouldn't we change from transferFrom function to transfer function for clarity?